shared/src/event_handling.rs

```diff
@@ -0,0 +1,181 @@
+use anyhow::{Context, Error, Result};
+use ethcontract::contract::{AllEventsBuilder, ParseLog};
+use ethcontract::errors::ExecutionError;
+use ethcontract::{dyns::DynTransport, BlockNumber as Web3BlockNumber, Event as EthcontractEvent};
+use futures::{Stream, StreamExt, TryStreamExt};
+use std::ops::RangeInclusive;
+use web3::Web3;
+
+// We expect that there is never a reorg that changes more than the last n blocks.
+const MAX_REORG_BLOCK_COUNT: u64 = 25;
+// When we insert new trade events into the database we will insert at most this many in one
+// transaction.
+const INSERT_EVENT_BATCH_SIZE: usize = 10_000;
+
+pub struct EventHandler<C: EventRetrieving, S: EventStoring<C::Event>> {
+    contract: C,
+    store: S,
+    last_handled_block: Option<u64>,
+}
+
+/// `EventStoring` is used by `EventHandler` for the purpose of giving the user freedom
+/// in how, where and which events are stored.
+///
+/// # Examples
+/// Databases: might transform, filter and classify which events are inserted
+/// HashSet: For less persistent (in memory) storing, insert events into a set.
+#[async_trait::async_trait]
+pub trait EventStoring<T> {
+    /// Returns ok, on successful execution, otherwise an appropriate error
+    ///
+    /// # Arguments
+    /// * `events` the contract events to be saved by the implementer
+    /// * `range` optionally indicates a particular range of blocks on which to operate.
```

Comment:
I find the information that if range is set we should consider all previously announced events as invalid (basically replace) a bit hidden here. It's very implicit in "if the Options is some, you need to replace".

Maybe we can either put back two methods (`replace_events(events, range)` vs `append_events(events)`) or consider leaving the splitting into `INSERT_EVENT_BATCH_SIZE` up to the event storing information and only call `save_events` once (in which case we could always provide the range) and document that the range might overlap in part with a previously announced event range.

Reply:
This might make it nice to make the two modes of operation for `EventStoring` more explicit.

Reply:
Yes, this is looking like a very nice change. Will update with this suggestion soon.